refactor: allow chain-specific configuration of Evm

Wodann opened the pull request:
Experiment to see whether it would be possible to separate optimism and mainnet `SpecId`s and other types, such as `HaltReason`.

I'm curious to hear whether any of this is in line with https://github.com/bluealloy/revm/issues/918

rakita reviewed (commented):
I have started reviewing this last weeks, but didnt have time to finish it whole and getter better grasp on it.

I general like it, change is needed but would like to minimise renames to get smaller PR.

As this is potentially a breaking change will need look at reth code and think how this impacts them as we are in the middle of devnet for Prague.

Wodann commented on crates/precompile/src/lib.rs:
```diff
@@ -236,8 +236,8 @@ pub enum PrecompileSpecId {
-    pub const fn from_spec_id(spec_id: revm_primitives::SpecId) -> Self {
-        use revm_primitives::SpecId::*;
+    pub const fn from_spec_id(spec_id: revm_primitives::EthSpecId) -> Self {
+        use revm_primitives::EthSpecId::*;
```
Resolved in https://github.com/bluealloy/revm/pull/1378/commits/c99bc9ae3e5392d43a2b2be3da904bb37ad2a3e2

Wodann commented on crates/primitives/src/chain_spec.rs:
```diff
@@ -0,0 +1,28 @@
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct MainnetChainSpec;
```
https://github.com/bluealloy/revm/pull/1378/commits/565eaa98060acc682dac6c882876df05a4043568

Wodann commented on crates/revm/benches/bench.rs:
```diff
@@ -10,8 +10,14 @@ use revm::{
+#[cfg(feature = "optimism")]
+type BenchChainSpec = revm::optimism::OptimismChainSpec;
+#[cfg(not(feature = "optimism"))]
+type BenchChainSpec = revm::MainnetChainSpec;
```
Resolved by https://github.com/bluealloy/revm/pull/1378/commits/53eb5b4948b043eee03a7dd1a2ec814158a0539d

Wodann commented on crates/revm/src/builder.rs:
```diff
@@ -28,61 +29,68 @@ pub struct SetGenericStage;
+impl<'a, ChainSpecT: ChainSpec, EXT, DB: Database>
+    EvmBuilder<'a, SetGenericStage, ChainSpecT, EXT, DB>
+{
+    pub fn with_chain_spec<NewChainSpecT: ChainSpec>(
```
Resolved by https://github.com/bluealloy/revm/pull/1378/commits/9d9e3e44b39fc83d404ed046c022289b615946db

Wodann commented on crates/revm/src/builder.rs:
```diff
@@ -28,61 +29,68 @@ pub struct SetGenericStage;
 pub struct HandlerStage;
 
-impl<'a> Default for EvmBuilder<'a, SetGenericStage, (), EmptyDB> {
+impl<'a, ChainSpecT: ChainSpec> Default
```
Resolved by https://github.com/bluealloy/revm/pull/1378/commits/5f87d2f5d7bd3399438e303bc58545a4eba060dd

Wodann commented on crates/primitives/src/chain_spec.rs:
```diff
@@ -0,0 +1,28 @@
+            type HaltReason: Clone + Debug + PartialEq + Eq + Hash + From<crate::HaltReason> + serde::de::DeserializeOwned + serde::Serialize;
+        } else {
+            /// The type that enumerates chain-specific halt reasons.
+            type HaltReason: Clone + Debug + PartialEq + Eq + Hash + From<crate::HaltReason>;
```
[This line](https://github.com/bluealloy/revm/pull/1378/files#diff-eba81aaaf8caad4f6f89c4e4f6da5ec5ec481c03e3ec32d1c36dbcbd53e27001R221) handles the conversion from `InstructionResult` to `type HaltReason` thanks to the `From<crate::HaltReason>`

Wodann commented on crates/primitives/src/chain_spec.rs:
```diff
@@ -0,0 +1,28 @@
+
+use core::{fmt::Debug, hash::Hash};
+
+pub trait ChainSpec: Clone + Copy + Debug + Sized + 'static {
```
Resolved by https://github.com/bluealloy/revm/pull/1378/commits/a09af7f407fbdfecf151a3c02fef456fd3baf138

Wodann commented on crates/primitives/src/result.rs:
```diff
@@ -1,26 +1,39 @@
 #[derive(Debug, Clone, PartialEq, Eq)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
-pub struct ResultAndState {
+pub struct ResultAndState<ChainSpecT: ChainSpec> {
```
The cast functions were a remnant from another attempt to achieve this. I've removed them in https://github.com/bluealloy/revm/pull/1378/commits/26dec59ba2c7af2207045088cbcf9bf3a5304d24.

W.r.t using the generic `ChainSpec` or `HaltReason`; up to you. I chose `HaltReason` in case some hardforks introduce custom `SuccessReason`s down the line.

Wodann commented on crates/revm/src/builder.rs:
```diff
@@ -28,61 +29,68 @@ pub struct SetGenericStage;
+        EvmBuilder {
+            context: self.context,
+            handler: EvmBuilder::<'_, SetGenericStage, NewChainSpecT, _, _>::handler(
+                NewChainSpecT::Hardfork::default(),
```
Is this something you want this PR to cover or a separate one?

rakita replied:
We can do that later. PR direction in general looks great!

Wodann commented on crates/precompile/Cargo.toml:
```diff
@@ -52,14 +52,6 @@ hashbrown = ["revm-primitives/hashbrown"]
-]
-negate-optimism-default-handler = [
-    "revm-primitives/negate-optimism-default-handler",
-]
```
If you're happy with the general approach, I can also adopt this pattern for the last couple of `#[cfg(feature = "optimism")] and then we can be certain that the features are gone from other parts of the code.

Wodann commented on crates/revm/src/custom_upcode.rs:
```diff
@@ -0,0 +1,253 @@
+use revm_interpreter::{
```
This file merely serves to illustrate that we can support custom opcodes. It should be removed before merging.

rakita replied:
We can move it to examples later, fine with leaving it here for now.

Wodann commented:
There are still some `EVMError::Custom` invocations in `handler_register.rs`. My understanding is that we could catch those when validating the environment.

That way we could use chain-specific errors static errors instead of strings. Alternatively, we could also add a custom error type to the `trait ChainSpec`.

That would allow us to remove the `EVMError::Custom(String)`

WDYT, @rakita ?

rakita commented on crates/revm/src/optimism/env.rs:
```diff
@@ -0,0 +1,176 @@
+    /// L1 block info used to compute the L1-cost fee.
+    ///
+    /// Needs to be provided for Optimism non-deposit transactions.
+    l1_block_info: Option<L1BlockInfo>,
```
Yeah, this looks nice

rakita commented:
I really like this path. For organizational purposes, lets merge this into chain_spec branch so we can work on it and make PR there, and after we are fine with the design, we can think about rebasing it on the main.

> I'll hold off on resolving merge conflicts until we're happy with the full design, if that's okay.

This is perfectly fine. 

> There are still some `EVMError::Custom` invocations in `handler_register.rs`. My understanding is that we could catch those when validating the environment.
> 
> That way we could use chain-specific errors static errors instead of strings. Alternatively, we could also add a custom error type to the `trait ChainSpec`.
> 
> That would allow us to remove the `EVMError::Custom(String)`

I would leave it as `String` for now, PR is getting big and we can add this later.

rakita commented:
I made a PR that refactors few things https://github.com/bluealloy/revm/pull/1528 I couldn't make it in nomicfoundation/revm didn't get the option for target repo to make a PR on.

I am still in investigation phase, but this looks great!

Wodann commented:
> I made a PR that refactors few things #1528 I couldn't make it in nomicfoundation/revm didn't get the option for target repo to make a PR on.
> 
> I am still in investigation phase, but this looks great!

Awesome! Is there anything you'd like me to do?

rakita commented:
> > I made a PR that refactors few things #1528 I couldn't make it in nomicfoundation/revm didn't get the option for target repo to make a PR on.
> > I am still in investigation phase, but this looks great!
> 
> Awesome! Is there anything you'd like me to do?

I dont have any action items rn, but I will ping you if something comes up.
PR is in good state, but it is a massive change in both the code and the design, so I want to be sure that it is okay. 

> I added your commits to this PR, squashed, and rebased on `main`

Great!

Wodann commented on crates/primitives/src/result.rs:
```diff
@@ -16,7 +16,6 @@ pub type EVMResultGeneric<T, ChainSpecT, DBError> = core::result::Result<
 >;
 
 #[derive(Debug, Clone, PartialEq, Eq)]
-#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
```
Removing serde (de)serialisation allowed the removal of the serde-related trait bounds from the `HaltReasonTrait`. As the `ExecutionResult` and `ResultAndState` aren't part of the JSON-RPC of Ethereum, I think it should be safe to remove this.

WDYT?

Wodann commented on crates/revm/src/context/evm_context.rs:
```diff
@@ -7,74 +8,55 @@ use crate::{
-{
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("EvmContext")
-            .field("inner", &self.inner)
-            .field("precompiles", &self.inner)
```
I used the `derive_where` procedural macro to avoid errors like this (double usage of `self.inner`).

Wodann commented on crates/revm/src/context/inner_evm_context.rs:
```diff
@@ -273,15 +246,10 @@ impl<DB: Database> InnerEvmContext<DB> {
+                // Use nonce from tx.
                 // Nonce for call is bumped in deduct_caller
                 // TODO(make this part of nonce increment code)
-                let nonce = self.env.tx.nonce.unwrap_or_else(|| {
-                    let caller = self.env.tx.caller;
-                    self.load_account(caller)
-                        .map(|(a, _)| a.info.nonce)
-                        .unwrap_or_default()
-                });
```
I moved the configuration of disabled nonce checks to the `CfgEnv` such that the transaction itself always requires a `nonce` value to be provided. If someone wants to disable nonce checks, they can do so for the whole `Evm`, similar to how it works for disabling other checks.

This voids the need for the `trait Transaction` to return an `Option<u64>` and I believe is in line with the overall design philosophy for having compile-time disable for checks like these.

Wodann commented on crates/primitives/src/transaction.rs:
```diff
@@ -0,0 +1,58 @@
+/// Trait for retrieving transaction information required for execution.
+pub trait Transaction {
+    /// Caller aka Author aka transaction signer.
+    fn caller(&self) -> &Address;
```
I used the philosophy that anything larger than a `u64` (pointer) should use a `&` whereas anything smaller or equal to a `u64` is returned by value. This is also the guideline for when to derive `Copy` or not, as it avoids unnecessarily large copies.

I'm not sure whether the presence of lifetimes in the return arguments might be a concern for some people though.

Wodann commented on crates/primitives/src/chain_spec.rs:
```diff
@@ -0,0 +1,57 @@
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct EthChainSpec;
+
+impl ChainSpec for EthChainSpec {
```
Would `L1ChainSpec` be a better/clearer name?

Wodann commented on crates/primitives/src/env.rs:
```diff
@@ -1,50 +1,47 @@
-    pub block: BlockEnv,
+    pub block: ChainSpecT::Block,
     /// Configuration of the transaction that is being executed.
-    pub tx: TxEnv,
```
As `BlockEnv` and `TxEnv` are no needed for the `EthChainSpec`, should they be moved to `chain_spec.rs`?

Wodann commented on crates/revm/src/builder.rs:
```diff
@@ -1,21 +1,22 @@
 /// will reset the registered handler to default mainnet.
-pub struct EvmBuilder<'a, BuilderStage, EXT, DB: Database> {
-    context: Context<EXT, DB>,
+pub struct EvmBuilder<'a, BuilderStage, ChainSpecT: ChainSpec, EXT, DB: Database> {
```
Instead of requiring `Default` for `ChainSpec::Transaction` and `ChainSpec::Block`, I think a nice design would be to have `EvmBuilder` contain `transaction: Option<ChainSpecT::Transaction>` and `block: Option<ChainSpecT::Block>`.

I feel that both should explicitly be set and when finishing the builder, a `Result` could be returned that constructs the `Context` or returns an error mentioning that no transaction and/or block was provided.

That way the user is not quietly using a default transaction or block configuration.

WDYT?

Wodann commented:
@rakita I rebased on `main` again and added some changes/fixes that I discovered after adopting the latest change into EDR.

Would it be a good idea to create a branch in `revm` with the `HEAD` of this branch, such that I can target that with additional (future) changes?

I also left some questions and notes to point out the changes that I made. I would appreciate if you could have a look at them and give me feedback.

Wodann commented:
I went through the PR diff and simplified the diff where possible.
